x/pairing/keeper/pairing.go

```diff
@@ -137,16 +140,42 @@ func (k Keeper) getPairingForClient(ctx sdk.Context, chainID string, clientAddre
 		return nil, 0, "", err
 	}
 
-	providers, err = k.calculatePairingForClient(ctx, possibleProviders, project.Index, block,
-		chainID, epochHash, strictestPolicy.MaxProvidersToPair)
+	// create the pairing slots with assigned reqs
+	slots := pairingscores.CalcSlots(strictestPolicy, minStake)
+
+	// group identical slots (in terms of reqs types)
+	slotGroups := pairingscores.GroupSlots(slots)
+
+	// create providerScore array with all possible providers
+	providerScores := []*pairingscores.PairingScore{}
+	for i := range possibleProviders {
+		providerScore := pairingscores.NewPairingScore(&possibleProviders[i])
+		providerScores = append(providerScores, providerScore)
+	}
+
+	// calculate score (always on the diff in score components of consecutive groups) and pick providers
+	prevGroupSlot := pairingscores.NewPairingSlot(map[string]pairingscores.ScoreReq{}) // init dummy slot to compare to
+	indexToSkipMapPtr := make(map[int]bool)                                            // keep the indices of chosen providers to we won't pick the same providers twice (for different groups)
+	for _, group := range slotGroups {
+		diffSlot := group.Slot.Diff(prevGroupSlot)
+		err := pairingscores.CalcPairingScore(providerScores, pairingscores.GetStrategy(), diffSlot, minStake)
+		if err != nil {
+			return nil, 0, "", err
+		}
+
+		pickedProviders := pairingscores.PickProviders(ctx, project.Index, providerScores, group.Count, block, chainID, epochHash, &indexToSkipMapPtr)
+		providers = append(providers, pickedProviders...)
```

Comment:
it looked like `PickProviders()` could pick more providers than asked. 
if that's the case, should you trim it before appending?

Reply:
PickProviders pick group.Count providers. All the groups' count add up to the amount of slots. The number of slots is always MaxProvidersToPair. So it wouldn't pick more